numpy/ctypeslib.py

```diff
@@ -213,6 +195,34 @@ def from_param(cls, obj):
         return obj.ctypes
 
 
+class _concrete_ndptr(_ndptr):
+    """
+    Like _ndptr, but with `_shape_` and `_dtype_` specified.
+
+    Notably, this means the pointer has enough information to reconstruct
+    the array, which is not generally true.
+    """
+    def _check_retval_(self):
+        """
+        This method is called when this class is used as the .restype
+        attribute for a shared-library function, to automatically wrap the
+        pointer into an array.
+        """
+        return self.contents
+
+    @property
+    def contents(self):
+        """
+        Get an ndarray viewing the data pointed to by this pointer.
+
+        This mirrors the `contents` attribute of a normal ctypes pointer
+        """
+        full_dtype = _dtype((self._dtype_, self._shape_))
+        full_ctype = ctypes.c_char * full_dtype.itemsize
+        buffer = ctypes.cast(self, ctypes.POINTER(full_ctype)).contents
+        return frombuffer(buffer, dtype=full_dtype).squeeze(axis=0)
```

Comment:
This is sort of frustrating - there ought to be a better way to construct an array from a pointer without having to go through an intermediate byte array

Reply:
Yeah, the goal of full_ctypes is to get an object supporting the buffer protocol that we can pass to `frombuffer`. Perhaps `ndarray.from_address(addr, shape, strides, dtype)` ought to be a (dangerous) part of the api?

Reply:
Interesting thought. I'd mostly be concerned if memory or speed was a problem with the current approach.

Reply:
Memory shouldn't be a concern - this doesn't make any copies. `.contents` returns a reference, not a copy.